types/egg/index.d.ts

```diff
@@ -0,0 +1,765 @@
+   *
+   * when you set locals, only object is available
+   *
+   * ```js
+   * this.locals = {
+   *   a: 1
+   * };
+   * this.locals = {
+   *   b: 1
+   * };
+   * this.locals.c = 1;
+   * console.log(this.locals);
+   * {
+   *   a: 1,
+   *   b: 1,
+   *   c: 1,
+   * };
+   * ```
+   *
+   * `ctx.locals` has cache, it only merges `app.locals` once in one request.
+   *
+   * @member {Object} Context#locals
+   */
+  locals: any;
+
+  /**
+   * alias to {@link locals}, compatible with koa that use this variable
+   */
+  state: any;
+
+  /**
+   * Logger for Application, wrapping app.coreLogger with context infomation
+   *
+   * @member {ContextLogger} Context#logger
+   * @since 1.0.0
+   * @example
+   * ```js
+   * this.logger.info('some request data: %j', this.request.body);
+   * this.logger.warn('WARNING!!!!');
+   * ```
+   */
+  logger: Logger;
+
+  /**
+   * Request start time
+   */
+  starttime: number;
+
+  /**
+   * View instance that is created every request
+   */
+  view: ContextView;
+
+  /**
+   * http request helper base on httpclient, it will auto save httpclient log.
+   * Keep the same api with httpclient.request(url, args).
+   * See https://github.com/node-modules/urllib#api-doc for more details.
+   */
+  curl(url: string, opt: any): Promise<any>;
+
+  /**
+   * Render a file by view engine
+   * @param {String} name - the file path based on root
+   * @param {Object} [locals] - data used by template
+   * @param {Object} [options] - view options, you can use `options.viewEngine` to specify view engine
+   * @return {Promise<String>} result - return a promise with a render result
+   */
+  render(name: string, locals: any, options?: any): Promise<string>;
+
+  /**
+   * Render a template string by view engine
+   * @param {String} tpl - template string
+   * @param {Object} [locals] - data used by template
+   * @param {Object} [options] - view options, you can use `options.viewEngine` to specify view engine
+   * @return {Promise<String>} result - return a promise with a render result
+   */
+  renderString(name: string, locals: any, options?: any): Promise<string>;
+
+  __(key: string, ...values: string[]): string;
+  gettext(key: string, ...values: string[]): string;
+
+  /**
+   * get upload file stream
+   * @example
+   * ```js
+   * const stream = yield this.getFileStream();
+   * // get other fields
+   * console.log(stream.fields);
+   * ```
+   * @method Context#getFileStream
+   * @return {ReadStream} stream
+   * @since 1.0.0
+   */
+  getFileStream(): Promise<FileStream>;
+
+  /**
+   * @see Responce.redirect
+   */
+  redirect(url: string, alt?: string): void;
+}
+
+export class Controller extends BaseContextClass { }
+
+export class Service extends BaseContextClass { }
+
+export interface service {} // tslint:disable-line
```

Comment:
Egg is a framework base on koa. And egg.js will injection service and controller to Application.

For example:

Here is a service at file `app/service/foo.js`
```js
import { Service } from 'egg';
class FooService extends Service {
   async bar() {
     return this.config.env;
   }
 }
```
Then service will inject to `ctx.service.foo` , so at controller can get the service this way:

file `app/controller/foo.js`
```js
import { Controller } from 'egg';

class FooController extends Controller {
  async getData() {
     this.ctx.body = await this.service.foo.bar();
   }
 }
```

So I define an empty interface service

```ts
export interface Context {
  service: service;
}
```

So I can add service property at the the service file

```ts
import { Service } from 'egg';
class FooService extends Service {
   async bar() {}
 }

declare module 'egg' {
   export interface service {
     foo: FooService;
   }
}
```

Then I can get service.foo at controller and other service file.